proposals/2020-07-01_aws-kms-keyring-redesign/change.md

```diff
@@ -0,0 +1,327 @@
+
+## Reference-level Explanation
+
+### AWS KMS symmetric keyring
+
+**Behavior**
+
+The AWS KMS symmetric keyring MUST...
+
+1. Be configured with a single string identifying an AWS KMS CMK
+   ([key name](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#key-names))
+   and MUST be configured with a single AWS SDK KMS service client (AWS SDK client) for that key name.
+2. Use the key name and AWS SDK client provided at initialization for encryption and decryption.
+3. Fail on decrypt if the encrypted data key’s provider info does not match the keyring’s configured key name.
+4. Provide the encrypted data key’s [provider info](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/structures.md#key-provider-information)
+   as part of the [AWS KMS Decrypt API](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html) call.
+
+The AWS KMS symmetric keyring MAY...
+
+1. Be configured with a list of string [grant tokens](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#grant-tokens)
+   to be included in all AWS KMS calls.
+
+**Decryption Contract**
+
+1. The keyring MUST have access to an AWS KMS client that is configured with credentials for an AWS principal
+   that passes an [AWS KMS Decrypt API](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html)
+   authorization check for the encrypted data key.
+2. The encrypted data key’s [provider info](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/structures.md#key-provider-information)
+   MUST match the keyring’s configured key name.
+3. The encryption context MUST exactly match the encryption context used on encrypt.
+
+**Initialization**
+
+On keyring initialization, an AWS KMS symmetric keyring MUST define the following:
+
+- Key Name
+- AWS SDK client
+
+On keyring initialization, an AWS KMS symmetric keyring MAY define the following:
+
+- Grant Tokens
+
+### AWS KMS symmetric region discovery keyring
+
+Customers MUST declare their intent to use the AWS KMS symmetric region discovery keyring instead of the AWS KMS symmetric keyring.
+
+**Behavior**
+
+The AWS KMS symmetric region discovery keyring MUST...
+
+1. Be configured with a single AWS SDK client.
+   1. Determine the AWS region associated with the provided AWS SDK client.
+      In some implementations,
+      this MAY require an additional AWS region argument
+      during AWS KMS symmetric region discovery keyring initialization.
+      If the AWS SDK returns the AWS SDK client’s region
+      directly from the AWS SDK client
+      in an intuitive manner (by using a single function/method call),
+      implementations MUST NOT require an additional AWS region argument.
```

Comment:
This will be tricky to strictly enforce, since it could easily be violated if a newer version of the SDK adds this "intuitive manner" under us. We'd have to set up a canary test case that asserts the exact maximum SDK version we depend on, so that it will break and force us to check on every version bump. It also implies a breaking change if this happens, since we then have to remove the parameter and force customers to stop providing it.

I'm not saying we shouldn't do this, but I'd feel better if we checked and found all current SDKs DO support this. Do we have hard evidence this currently isn't possible in any of them? Should we push for all SDKs to?

Reply:
My understanding is that the Java AWS SDK makes it difficult to obtain the AWS region *after* the AWS SDK client has been initialized.

This wouldn't necessarily be a breaking change. From an implementation standpoint this would likely be...
```
Before:
MakeKeyring(sdkClient, region):
  // Do stuff


After:
MakeKeyring(sdkClient):
  region = GetRegion(sdkClient);
  MakeKeyring(sdkClient, region);

MakeKeyring(sdkClient, region):
  // Do Stuff
```

`MUST NOT require` Does not imply `MAY not provide`, so supporting both is still upholding this behavior.